File: CS/XPO/RuntimeDbChooser.Win/Controllers/CustomLogonController.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using RuntimeDbChooser.Module.BusinessObjects;

namespace RuntimeDbChooser.Win.Controllers {
    public class CustomLogonController : LogonController {
        protected override void Accept(SimpleActionExecuteEventArgs args) {
            string targetDataBaseName = ((IDatabaseNameParameter)Application.Security.LogonParameters).DatabaseName;
            string newConnectionString = MSSqlServerChangeDatabaseHelper.PatchConnectionString(targetDataBaseName, Application.ConnectionString);
            if(Application.ConnectionString != newConnectionString) {
                Application.ObjectSpaceProviderContainer?.Clear();
                Application.ConnectionString = newConnectionString;
            }
            base.Accept(args);
        }
    }
}

File: VB/XPO/NET_Framework/RuntimeDbChooser.Win/Controllers/CustomLogonController.vb
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports RuntimeDbChooser.Module.BusinessObjects

Namespace RuntimeDbChooser.Win.Controllers

    Public Class CustomLogonController
        Inherits LogonController

        Protected Overrides Sub Accept(ByVal args As SimpleActionExecuteEventArgs)
            Dim targetDataBaseName As String = CType(Application.Security.LogonParameters, IDatabaseNameParameter).DatabaseName
            Dim newConnectionString As String = MSSqlServerChangeDatabaseHelper.PatchConnectionString(targetDataBaseName, Application.ConnectionString)
            If Not Equals(Application.ConnectionString, newConnectionString) Then
                Application.ObjectSpaceProviderContainer?.Clear()
                Application.ConnectionString = newConnectionString
            End If

            MyBase.Accept(args)
        End Sub
    End Class
End Namespace
